pkg/devfile/validate/events.go

```diff
@@ -0,0 +1,96 @@
+package validate
+
+import (
+	"fmt"
+	"strings"
+
+	adapterCommon "github.com/openshift/odo/pkg/devfile/adapters/common"
+	"github.com/openshift/odo/pkg/devfile/parser/data"
+	"k8s.io/klog"
+)
+
+// validateEvents validates all the devfile events
+func validateEvents(data data.DevfileData) error {
+
+	eventErrors := ""
+	events := data.GetEvents()
+
+	switch {
+	case len(events.PreStart) > 0:
+		klog.V(4).Info("Validating preStart events")
+		if preStartErr := isEventValid(data, events.PreStart, "preStart"); preStartErr != nil {
+			eventErrors += fmt.Sprintf("\n%s", preStartErr.Error())
+		}
+		fallthrough
+	case len(events.PostStart) > 0:
+		klog.V(4).Info("Validating postStart events")
+		if postStartErr := isEventValid(data, events.PostStart, "postStart"); postStartErr != nil {
+			eventErrors += fmt.Sprintf("\n%s", postStartErr.Error())
+		}
+		fallthrough
+	case len(events.PreStop) > 0:
+		klog.V(4).Info("Validating preStop events")
+		if preStopErr := isEventValid(data, events.PreStop, "preStop"); preStopErr != nil {
+			eventErrors += fmt.Sprintf("\n%s", preStopErr.Error())
+		}
+		fallthrough
+	case len(events.PostStop) > 0:
+		klog.V(4).Info("Validating postStop events")
+		if postStopErr := isEventValid(data, events.PostStop, "postStop"); postStopErr != nil {
+			eventErrors += fmt.Sprintf("\n%s", postStopErr.Error())
+		}
+	}
+
+	// if there is any validation error, return it
+	if len(eventErrors) > 0 {
+		return fmt.Errorf("devfile events validation error: %s", eventErrors)
+	}
+
+	return nil
+}
+
+// isEventValid checks if events belonging to a specific event type are valid:
+// 1. Event should map to a valid devfile command
+// 2. Event commands should be valid
+func isEventValid(data data.DevfileData, eventNames []string, eventType string) error {
+	eventErrorMsg := make(map[string][]string)
+	eventErrors := ""
+	commands := data.GetCommands()
+
+	for _, eventName := range eventNames {
+		isEventPresent := false
+		for _, command := range commands {
+			// Check if event matches a valid devfile command
+			if command.GetID() == strings.ToLower(eventName) {
+				isEventPresent = true
+
+				// Check if the devfile command is valid
+				err := adapterCommon.ValidateCommand(data, command)
```

Comment:
commands should be validated separately(not saying to resolve in this PR)

Reply:
How would you propose command to be validated that does not involve `odo push`? Here I am reusing the func we use to validate commands for `odo push`

Reply:
all devfile commands should be validated just after parsing, like we are doing for `events`,  if all commands validated before only, no need to validate them here.

Reply:
https://github.com/openshift/odo/pull/3631/files#diff-838aff5d94d44c245e0db507c96219d9R24

Reply:
Agreed, opened an issue to refactor it https://github.com/openshift/odo/issues/3703